pkg/sidecar-controller/groupsnapshot_helper.go

```diff
@@ -159,12 +167,477 @@ func (ctrl *csiSnapshotSideCarController) syncGroupSnapshotContent(content *crdv
+		parameters[utils.PrefixedVolumeGroupSnapshotContentNameKey] = content.Name
+	}
+
+	volumeIDs, err := ctrl.getGroupSnapshotVolumeIDs(content)
+	driverName, groupSnapshotID, snapshots, creationTime, readyToUse, err := ctrl.handler.CreateGroupSnapshot(content, volumeIDs, parameters, snapshotterCredentials)
+	if err != nil {
+		// NOTE(xyang): handle create timeout
+		// If it is a final error, remove annotation to indicate
+		// storage system has responded with an error
+		klog.Infof("createSnapshotWrapper: CreateSnapshot for content %s returned error: %v", content.Name, err)
+		if isCSIFinalError(err) {
+			var removeAnnotationErr error
+			if content, removeAnnotationErr = ctrl.removeAnnVolumeGroupSnapshotBeingCreated(content); removeAnnotationErr != nil {
+				return content, fmt.Errorf("failed to remove VolumeGroupSnapshotBeingCreated annotation from the content %s: %s", content.Name, removeAnnotationErr)
+			}
+		}
+
+		return content, fmt.Errorf("failed to take group snapshot of the volumes %s: %q", content.Spec.Source.PersistentVolumeNames, err)
+	}
+
+	klog.V(5).Infof("Created group snapshot: driver %s, groupSnapshotId %s, creationTime %v, readyToUse %t", driverName, groupSnapshotID, creationTime, readyToUse)
+
+	if creationTime.IsZero() {
+		creationTime = time.Now()
+	}
+
+	// Create individual snapshots and snapshot contents
+	for _, snapshot := range snapshots {
+		volumeSnapshotContentName := GetSnapshotContentNameForVolumeGroupSnapshotContent(content)
+		volumeSnapshotName := GetSnapshotNameForVolumeGroupSnapshotContent(content)
+		volumeSnapshotNamespace := content.Spec.VolumeGroupSnapshotRef.Namespace
+		volumeSnapshotContent := &crdv1.VolumeSnapshotContent{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: volumeSnapshotContentName,
+			},
+			Spec: crdv1.VolumeSnapshotContentSpec{
+				VolumeSnapshotRef: v1.ObjectReference{
+					Kind:      "VolumeSnapshots",
+					Name:      volumeSnapshotName,
+					Namespace: volumeSnapshotNamespace,
+				},
+				DeletionPolicy: content.Spec.DeletionPolicy,
+				Driver:         content.Spec.Driver,
+				Source: crdv1.VolumeSnapshotContentSource{
+					SnapshotHandle: &snapshot.SnapshotId,
+				},
+				SourceVolumeMode: nil,
+			},
+		}
+
+		volumeSnapshot := &crdv1.VolumeSnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      volumeSnapshotName,
+				Namespace: volumeSnapshotNamespace,
+			},
+			Spec: crdv1.VolumeSnapshotSpec{
+				Source: crdv1.VolumeSnapshotSource{
+					VolumeSnapshotContentName: &volumeSnapshotContentName,
+				},
+			},
+		}
+		_, err = ctrl.clientset.SnapshotV1().VolumeSnapshotContents().Create(context.TODO(), volumeSnapshotContent, metav1.CreateOptions{})
+		if err != nil {
+			return content, err
+		}
+
+		_, err = ctrl.clientset.SnapshotV1().VolumeSnapshots(volumeSnapshotNamespace).Create(context.TODO(), volumeSnapshot, metav1.CreateOptions{})
+		if err != nil {
+			return content, err
+		}
+
```

Comment:
Where do we updateGroupSnapshotStatus?  We need to have `VolumeSnapshotRefList []core_v1.ObjectReference` in GroupSnapshotStatus.

Reply:
Yeah that's a little trickier, since we don't have a reference to the individual volume snapshots at the point of updating the status. Ill take a look at how to modify the logic..

Reply:
Done in this commit - https://github.com/kubernetes-csi/external-snapshotter/pull/826/commits/644db3b80954f3da88395a4302d5d4d5ce960088

```
% kubectl get vs
NAME                                                                                           READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT                                                                             RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT                                                                                   CREATIONTIME   AGE
snapshot-130580a5d5a1d00b87b486cb2c606f2f8edad26bc804dbb934c9dd16747afb3f-2023-04-28-6.51.19   true                     snapcontent-130580a5d5a1d00b87b486cb2c606f2f8edad26bc804dbb934c9dd16747afb3f-2023-04-28-6.51.19   1Gi                           snapcontent-130580a5d5a1d00b87b486cb2c606f2f8edad26bc804dbb934c9dd16747afb3f-2023-04-28-6.51.19   2m45s          2m45s
snapshot-a536d183ab7e835ef0970c5d4dc1e04e8b8e0e3a3ab4b6c21b08465994a2e600-2023-04-28-6.51.19   true                     snapcontent-a536d183ab7e835ef0970c5d4dc1e04e8b8e0e3a3ab4b6c21b08465994a2e600-2023-04-28-6.51.19   1Gi                           snapcontent-a536d183ab7e835ef0970c5d4dc1e04e8b8e0e3a3ab4b6c21b08465994a2e600-2023-04-28-6.51.19   2m45s          2m45s

% kubectl describe vgs
Name:         my-group-snapshot
Namespace:    default
...
Status:
  Bound Volume Group Snapshot Content Name:  groupsnapcontent-7f289eac-31b2-4980-9bee-fa0402e52cea
  Creation Time:                             2023-04-28T06:51:19Z
  Ready To Use:                              true
  Volume Snapshot Ref List:
    Kind:       VolumeSnapshots
    Name:       snapshot-a536d183ab7e835ef0970c5d4dc1e04e8b8e0e3a3ab4b6c21b08465994a2e600-2023-04-28-6.51.19
    Namespace:  default
    UID:        66ea779d-1c21-4331-9083-5d9766c3a706
    Kind:       VolumeSnapshots
    Name:       snapshot-130580a5d5a1d00b87b486cb2c606f2f8edad26bc804dbb934c9dd16747afb3f-2023-04-28-6.51.19
    Namespace:  default
    UID:        094a1a03-90ea-408a-aa80-8cd6897d2f7a
```

Reply:
Can you update the PR description section with the new testing output?